cre.cpp

```diff
@@ -1392,6 +1392,48 @@ static int isXPointerInCurrentPage(lua_State *L) {
 	return 1;
 }
 
+static int hasXPointer(lua_State *L) {
+    CreDocument *doc = (CreDocument*) luaL_checkudata(L, 1, "credocument");
+    const char *xpointer_str = luaL_checkstring(L, 2);
+
+    bool found = true;
+    ldomXPointer xp = doc->dom_doc->createXPointer(lString16(xpointer_str));
+    lvPoint pt = xp.toPoint();
+    if (pt.y < 0) {
+        found = false;
+    }
+    lua_pushboolean(L, found);
+    return 1;
+}
+
+static int getImageDataFromPosition(lua_State *L) {
+    CreDocument *doc = (CreDocument*) luaL_checkudata(L, 1, "credocument");
+    int x = luaL_checkint(L, 2);
+    int y = luaL_checkint(L, 3);
+    lvPoint pt(x, y);
+    ldomXPointer ptr = doc->text_view->getNodeByPoint(pt);
+    if (ptr.isNull())
+        return 0;
+    LVStreamRef stream = ptr.getNode()->getObjectImageStream();
+    if (!stream.isNull()) {
+        unsigned size = stream->GetSize();
+        lvsize_t read_size = 0;
+        void *buffer = (void *)malloc(size);
```

Comment:
It's generally not a good practice to mix LUA C binding with LUAJIT FFI. We had to use C binding for crengine because LUAJIT does not play well with C++.

The proper way to do memory management in LUA C binding is to use metatable with userdata. lightuserdata is mostly used when you need to do equality comparison, see: https://www.lua.org/pil/28.5.html.

Basically you will need to create a new metatable for the image stream in luaopen_cre. Then sets the `__gc` method in the metatable. Then set this metatable to your image stream userdata in getImageDataFromPosition. You can find an example in newDocView where it does the same thing for CreDocument. We set both `__gc` and `close` methods for CreDocument to the same function so it can be freed either manually or automatically by LuaJIT's GC.